### DISCUSSION TITLE
How to insert this formula correct?

### DISCUSSION THREAD

**Discussion by monoeagle** [dfinke/ImportExcel#1637] [Q&A] [answered]
Hi,

the following AllData tab is the source.
![AllDataPage](https://github.com/user-attachments/assets/9424356f-e9e4-44ba-9301-f6cc4480a012)

On that tab I want to build an dynamic table for the diagram.
![PageActionsPerYear](https://github.com/user-attachments/assets/8f2b8d2e-152a-497b-a412-e9a30ef8b2be)

If I use the formular manually it works:

`
=SUMME(BEREICH.VERSCHIEBEN(AllData!$A$2:$A$128;0;VERGLEICH("o_2020";AllData!$A$1:$AA$1;0)-1))
`

But if I add this in the formula property of set-excelrange I allways get an error if I try to open the generated excel file with the message remove incorrect parts and the cells are empty.

`
 Set-ExcelRange -Address $ActionsPerYearWorksheet.Cells["B2"] -Formula "=SUMME(BEREICH.VERSCHIEBEN(AllData!$A$2:$A$128;0;VERGLEICH(o_2020;AllData!$A$1:$AA$1;0)-1))"
`

I tried to escape "$A$2:$A$128" and "$A$1:$AA$1" with ` or make a string out of it $(" ") but without success.
Excel also didn't tell me what it doesn't like.

Is there a special form for the formula property?

greetz

**Comment by dfinke** [comment]
Yes, that is what I would've suggested. Excel just does not do anything with it?

Would need to see code, data, and xlsx.

**Reply by monoeagle** [reply]
Here we are. =)

[Data.csv](https://github.com/user-attachments/files/17094915/Data.csv)
[script.txt](https://github.com/user-attachments/files/17094947/script.txt)
[Stat.xlsx](https://github.com/user-attachments/files/17094951/Stat.xlsx)

**Comment by dfinke** [comment]
sorry - the data etc are fine. need a simple script that repros the problem.

**Reply by monoeagle** [reply]
Hmm, that is the script with the problem. And and I can reproduce the error here with that script.

The Syntax in the script and the escaping is ok?
If I try to open the generated xlsx File I get the following message:

![error](https://github.com/user-attachments/assets/007c3a3e-344f-44dc-a463-1d984024678f)

If I press Yes the formulas will be removed.

Could it be an excel issue?
We use Office 2019 Std (1808).

**Reply by monoeagle** [reply]
![Error](https://github.com/user-attachments/assets/8519a0cf-2e4d-4e06-ae5c-1090c6f3022f)

If I use ' at the beginning and at the end vom the formula the excel file will open without an error with hardcoded o_2020; m_2020 and c_2020. And the translation to gern will be automatically if I write it in english in the script.

`
=SUMME(BEREICH.VERSCHIEBEN(AllData!$A$2:$A$128;0;VERGLEICH("m_2020";AllData!$A$1:$AA$1;0)-1))
`

But if I try to insert the $yr Variable in the Match it doesn't work.

So the problem is in the escaping.

**Reply by dfinke** [reply]
### Maybe
```
`"o_$yr`"
```

**Reply by monoeagle** [reply]
Nope that is what I tried in the uploaded script.txt 2 days ago.

**Reply by dfinke** [reply]
i would try, build the parts that work. don't insert that formula with o_$yr. print out that string to the console, copy paste it to Excel. get more clues

**Reply by monoeagle** [reply]
crazy behavior
manual dummy table with one sheet for the data and a second sheet for the results
![20240926_AllData Dummy Table](https://github.com/user-attachments/assets/1e6b0fe7-c659-4608-8738-bc8c49410b9c)

In the column "C" I wrote the formula text which I used in Column "B"
![20240926_Result from Dummy Table](https://github.com/user-attachments/assets/312242e9-b75c-4e69-bf21-c3d58c80c38d)
![20240926_Formula in german](https://github.com/user-attachments/assets/23417038-435b-4714-a8b7-6419b7bf35ba)

The result of the manual try in the sheet2.xml is on the right side.

The german formula is translated in english automatically. 

[PICTURE DIFF]
![Diff Sheet2](https://github.com/user-attachments/assets/55ebb4c3-8228-4443-a8a9-f440761dd85a)

But if I use the translated formula manual in the cell B4 it results in #NAME? and in the xml file  
`_xludf.SUM(_xludf.OFFSET(AllData!$A$2:$A$128,0,_xludf.MATCH("m_2020",AllData!$A$1:$AA$1,0)-1))
`

He didn't know SUM, OFFSET and MATCH.

The left side of the [PICTURE DIFF] is the generated xlsx file via the import-excel module.
![script](https://github.com/user-attachments/assets/ea37bade-6145-45f6-a261-fdbf8c65fe12)
In the script I use the escaped "o_$yr" and in the xml file we can see the result is as expected.
But Excel says that there is a problem.

If I use the formula in german with excaping:
![script_german](https://github.com/user-attachments/assets/c60028be-58bd-456f-a4df-a7bd2b059038)

The result in the XML on the left side
![20240926_DIFF_not translated](https://github.com/user-attachments/assets/852d60df-61b1-4712-9fe2-c7247d3228f4)

The escaping works, but excel already has a problem.

Without the $yr variable --> same error

![script_english_without yr variable](https://github.com/user-attachments/assets/90e88449-9c70-4499-93e4-882fd5c427d7)

In the XML File like in [PICTURE DIFF] on the left side the formula is correct but excel want to repair/remove them.

It's crazy!

**Reply by dfinke** [reply]
Definitely strange. I don't know.

**Reply by monoeagle** [reply]
I'm digging. ;)

**Comment by monoeagle** [comment, marked as answer]
Problem Solved!
![20241001_formelauflösung](https://github.com/user-attachments/assets/2a7424ed-57ed-4bd8-92f4-96d5114c8ae3)

In General where I had to look at:
- in german the delemiter is ";" in englisch ","
- to build the formula seprately before add it to "-Fomula"

Result is as wanted:
![20241001_Ergebnis](https://github.com/user-attachments/assets/cbd8ec6a-99e7-4e67-aa09-12099b984936)

**Comment by dfinke** [comment]
Wow - so  "lanuage pack" influenced this?

**Reply by monoeagle** [reply]
It looks like that. 
Writing german formula in excel manually excel will translate them automatically to write them into the xml. ;)
So it is better to use the english version with generated xlsx. ;)

But also how to write the formula especially if you use variables in the formula.
Put the Formula in a separate variable works for me.
